### PR TITLE
docs: provide minimal init.lua and recommend it

### PR DESCRIPTION
**Pull request by clason**
Add a `minimal.lua` for reproducing issues in a reproducible clean
environment (config _and_ plugins/packages). This bootstraps packer into
`/tmp/nvim` (so unixy platforms only for now; Windows users need to
manually change the paths to `%TEMP%`) if needed, synchronizes all
specified plugins (to ensure up-to-date environment even if used
previously) and loads the specified `init.lua` settings.

Mention this in `CONTRIBUTING.md` and the bug issue template.

@dundargoc @vigoux

### CODE REVIEW

**Comment by jamessan** [comment]
`ci` seems like the wrong type for this commit. I've noticed that being used for other commits related to the templates, too.  Since there's no `meta` type, `chore` looks like the best fit.

Signed,
Still skeptical about conventional commits :-)

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 1]
```diff
@@ -43,8 +37,8 @@ body:
     attributes:
       label: "Steps to reproduce"
       description: |
-        Steps to reproduce using `nvim -u NORC` and/or `nvim -u NONE` (please test both).
-        If you are reporting build failures, please list the exact sequence of steps including all CMake flags (if any).
+        If you cannot reproduce the issue using `nvim -u NORC`, prepare and upload a [minimal config](https://github.com/neovim/neovim/blob/master/contrib/minimal.lua) that reproduces the issue with `nvim -u minimal.lua`.
```
Comment (dundargoc) [inline .github/ISSUE_TEMPLATE/bug_report.yml:40]: Upload where? In a gist or paste it in the report?

Comment (dundargoc) [inline .github/ISSUE_TEMPLATE/bug_report.yml:40]: Upload where? In a gist or paste it in the report?

Comment (dundargoc) [inline .github/ISSUE_TEMPLATE/bug_report.yml:40]: It might be obvious to us, but I think we could benefit from expanding on this a little. Reminding the issuer that the smaller the configuration is, the more likely we'll be able to pinpoint the problem and the more likely we'll be able to resolve their issue ("help us help you"). I know this is already implied by the "minimal" part but my hope is that expressing that we _want_ to solve their issues but need their help might encourage better reports.

File: contrib/minimal.lua [hunk 2]
```diff
@@ -0,0 +1,30 @@
+vim.cmd [[set runtimepath=$VIMRUNTIME]]
```
Comment (dundargoc) [inline contrib/minimal.lua:1]: Don't know how to review an entire file so picked the first line.

Don't we also need a template for people who use init.vim? If not then maybe we should mention "If you use init.lua use the minimal.lua template, otherwise just minimize your current config." .

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: That _would_ be useful to explain, but probably not here. We could add a section to `CONTRIBUTING.md` and link that?

File: contrib/minimal.lua (hunk 2 above)
Comment (clason) [reply contrib/minimal.lua:1]: I think there's only a tiny subset of issues that _require_ an `init.vim` (as in, cannot be reproduced  with an `init.lua`). And vimscript config can be used in the config block by wrapping it in `vim.cmd[[ ... ]]`.

Could be explained more, of course, but again, maybe not here?

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (dundargoc) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: Hmm yeah, probably a bit too wordy. I don't think we need to link it honestly, it'll just add noise. I'm backpedaling on this one, think we can drop it.

Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: I'm open for suggestions how best to share the customized template :)

Comment (dundargoc) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: What? I need to have permission to resolve _my own_ reviews?

Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: No, if we _do_ add such a section, it could replace the link to the file. It's something I'll think about.

Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: (Probably just a conflict because we commented at the same time.)

Comment (dundargoc) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: > Good question; I flipped-flopped on that as well, hence it's not mentioned :) Unfortunately, `.lua` is not a supported filetype for uploading...

For uploading where? In gists?

Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: By attaching via drag & drop to the issue itself.

Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:40]: Yes, there's only a (somewhat eclectic) list of allowed filetypes.

I don't like the spoiler tag; if it's truly minimal, just copying the _added_ lines in the "steps" field would work.

**Comment by dundargoc** [comment]
I don't know how to approve pull requests but I'll do it here: :+1:

File: contrib/minimal.lua (hunk 2 above)
Comment (dundargoc) [reply contrib/minimal.lua:1]: You may resolve this conversation if you'd like. I don't have the permission to do it myself.

Comment (dundargoc) [reply contrib/minimal.lua:1]: Same goes for the other conversations really.

**Review by justinmk — APPROVED** [review]

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 3]
```diff
@@ -52,24 +51,33 @@ body:
...
     validations:
       required: true
+
+  - type: textarea
+    attributes:
+      label: "Minimal config"
+      description: "Minimal(!) configuration necessary to reproduce the issue. You can start with the provided [`minimal.lua`](https://github.com/neovim/neovim/blob/master/contrib/minimal.lua). If _absolutely_ necessary, add plugins and config options from your `init.lua` at the indicated lines."
+      render: Lua
+    validations:
+      required: true
```
Comment (justinmk) [inline .github/ISSUE_TEMPLATE/bug_report.yml:82]: This is quite some friction to add to bug reports. No objection to having and mentioning the minimal.lua, but making it a required part of bug reports now means that it basically replaces `nvim -u NONE` or `nvim -u NORC` for both bug-reporters and bug-triagers. That is going to be mostly noise for most reports.

**Review by justinmk — CHANGES_REQUESTED** [review]
not in favor of making this required

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 4]
```diff
@@ -52,24 +51,33 @@ body:
...
+        - [Prepare](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#reporting-problems) a minimal config that reproduces the problem with `nvim -u minimal.lua` and provide it below.
         - For build failures: list the exact steps including CMake flags (if any).
         - For shell-related problems: try `env -i TERM=ansi-256color "$(which nvim)"`.
       placeholder: |
-        nvim -u NONE
-        :edit foo
-        yiwp
+        1. `nvim -u minimal.lua` 
+        2. `:edit foo` 
+        3. `yiwp`
```
Comment (justinmk) [inline .github/ISSUE_TEMPLATE/bug_report.yml:60]: this is harder to type for the user, and not really more readable for triagers. and it makes copying the text with a mouse, annoying.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -7,24 +7,31 @@ Getting started
...
 Reporting problems
 ------------------
 
+Before opening a new issue:
+
 - [Check the FAQ][wiki-faq].
 - [Search existing issues][github-issues] (including closed!)
 - Update Neovim to the latest version to see if your problem persists.
-- [Bisect](https://neovim.io/doc/user/starting.html#bisect) your config: disable plugins incrementally, to narrow down the cause of the issue.
-- [Bisect][git-bisect] Neovim's source code to find the cause of a regression, if you can. This is _extremely_ helpful.
```
Comment (justinmk) [inline CONTRIBUTING.md:23]: this form was intentional as a parallel with the other "bisect" case.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -7,24 +7,31 @@ Getting started
...
+
+When reporting an issue:
+
+- [Bisect](https://neovim.io/doc/user/starting.html#bisect) your config: disable plugins and remove settings incrementally, to narrow down the cause of the issue.
+- Using this information, prepare a [minimal config](https://github.com/neovim/neovim/blob/master/contrib/minimal.lua) that reproduces the problem with `nvim -u minimal.lua`.
```
Comment (justinmk) [inline CONTRIBUTING.md:28]: this is too much. we can weakly suggest it, but to make it now the standard for bug reports means that everyone has to have this thing hanging around. most bug reports don't need this.

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 3 above)
Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:82]: Here's the basic problem this is supposed to solve: with the advent of `packer.nvim`, vim packages have become a de facto standard for plugins. And neither `-u NONE` nor `-u NORC` is suitable for dealing with those:

1. `NORC` still sources all `start` packages.
2. `NONE` (and `--noplugins`) skip _all_ runtime files -- even the builtin ones.

This has lead to a lot of "bug" reports that are actually conflicts with installed plugins.

This init file is the only way I have found to have a reliably and reproducibly clean state.

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 4 above)
Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:60]: It's purely cosmetic for the placeholder?

File: CONTRIBUTING.md (hunk 6 above)
Comment (clason) [reply CONTRIBUTING.md:28]: The second motivation for this change is that leaving this open leads to many reports that just skip these steps even though they are needed for diagnosing the issue, forcing maintainers to request additional information.

**Comment by clason** [comment]
To be blunt, this is becoming a bit much of a moving target for me; we clearly have incompatible views about some of this, and I don't see much chance of a quick compromise.

I will therefore step back from this PR and let you make any changes you deem necessary, or close it. (No hard feelings either way.)

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 3 above)
Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:82]: And, for the record: The reason to have this as a separate field is to

1. not clutter the "steps" further and
2. allow marking this as a code block so you get a) syntax highlighting and b) a button for copying the config.

This reduces friction for maintainers trying to reproduce the issue.

Comment (justinmk) [reply .github/ISSUE_TEMPLATE/bug_report.yml:82]: It doesn't really reduce friction. I almost never install plugins to repro an issue, and I don't want to start doing so. I'm also imagining a veteran vim user trying to file a bug report and getting hit with these bureaucratic steps, and then just giving us the middle finger...

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 4 above)
Comment (justinmk) [reply .github/ISSUE_TEMPLATE/bug_report.yml:60]: safari bugs (temporary) don't dictate our issue template. it's not purely cosmetic, because this establishes the pattern that many will follow or feel urged to follow.

File: CONTRIBUTING.md (hunk 6 above)
Comment (justinmk) [reply CONTRIBUTING.md:28]: needing to install plugins should be rare for repro steps. huge burden for triagers. it is more useful if the repot narrows down the steps to actual code.

we should _rarely_ be debugging plugins, that massively increases the triage scope.

**Comment by justinmk** [comment]
> 1. `NORC` still sources all `start` packages.
> 2. `NONE` (and `--noplugins`) skip _all_ runtime files -- even the builtin ones.
> 
> This has lead to a lot of "bug" reports that are actually conflicts with installed plugins.

sounds like we are converging on the _actual_ problem. (does `--clean` help at all, btw?)

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 3 above)
Comment (clason) [reply .github/ISSUE_TEMPLATE/bug_report.yml:82]: > sounds like we are converging on the _actual_ problem. (does `--clean` help at all, btw?)

No, same issue (`clean` just additionally skips `shada`).

File: CONTRIBUTING.md (hunk 6 above)
Comment (clason) [reply CONTRIBUTING.md:28]: Agreed, but this makes it _immediately_ obvious that (and which) plugins are involved instead of having to pull that information like a tooth.

**Comment by clason** [comment]
Now that `--clean` behaves as intended, recommend that over the minimal `init.lua`. (Might still be useful so I'm keeping it in the PR; but can be dropped if that is preferable.)

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -7,24 +7,32 @@ Getting started
...
+
+When reporting an issue:
+
+- First try to reproduce with `nvim --clean` ("factory defaults").
+- [Bisect](https://neovim.io/doc/user/starting.html#bisect) your config: disable plugins and remove settings incrementally, to narrow down the cause of the issue.
+- If the problem involves plugins, preparing a [minimal config](https://github.com/neovim/neovim/blob/master/contrib/minimal.lua) that reproduces the problem with `nvim -u minimal.lua` can be very helpful.
+- Check the logs: `:edit $NVIM_LOG_FILE`.
+- For shell-related problems: try `env -i TERM=ansi-256color "$(which nvim)"`.
+- Include `cmake --system-information` for build-related issues.
+- If there issue is a regression (i.e., the problem did not occur in an earlier version), [bisect][git-bisect] Neovim's source code to find exactly when it was introduced. This is _extremely_ helpful.
```
Comment (justinmk) [inline CONTRIBUTING.md:33]: typos. also the new, more wording is more verbose without really gaining anything
